### PR TITLE
Add a path deviation analytics block to workflows

### PR DESCRIPTION
**Pull request by shantanubala**
# Description

This PR includes an analytics block that can be useful for analyzing the path or trajectory of detected objects in an image. It implements an algorithm for computing the [Fréchet distance](https://en.wikipedia.org/wiki/Fr%C3%A9chet_distance) between the path of a detected object and an optimal path.

> Imagine a person traversing a finite curved path while walking their dog on a leash, with the dog traversing a separate finite curved path. Each can vary their speed to keep slack in the leash, but neither can move backwards. The Fréchet distance between the two curves is the length of the shortest leash sufficient for both to traverse their separate paths from start to finish.

This can be useful for tasks like counting the number of cars taking a left turn, measuring the precision of a baseball pitch, or any other task where you want to measure the extent to which an object deviates from a desired path.

## Type of change

Please delete options that are not relevant.

-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

TBD.

## Any specific deployment considerations

No specific deployment considerations - it is a workflow block that builds on top of the existing structure.

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

**Review by PawelPeczek-Roboflow — DISMISSED** [review]

File: inference/core/workflows/core_steps/analytics/line_following/v1.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        ] = {}
+
+    @classmethod
+    def get_manifest(cls) -> Type[WorkflowBlockManifest]:
+        return LineFollowingManifest
+
+    def _calculate_frechet_distance(
+        self, path1: np.ndarray, path2: np.ndarray
+    ) -> float:
+        def euclidean_distance(point1, point2):
```
[the remark this answers is not in the document]
Comment (shantanubala) [reply inference/core/workflows/core_steps/analytics/line_following/v1.py:96]: Just fixed 👍

File: inference/core/workflows/core_steps/analytics/line_following/v1.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    def __init__(self):
+        self._object_paths: Dict[
+            str, Dict[Union[int, str], List[Tuple[float, float]]]
+        ] = {}
+
+    @classmethod
+    def get_manifest(cls) -> Type[WorkflowBlockManifest]:
+        return LineFollowingManifest
+
+    def _calculate_frechet_distance(
```
[the remark this answers is not in the document]
Comment (shantanubala) [reply inference/core/workflows/core_steps/analytics/line_following/v1.py]: Thank you for the quick review, updated this as well 🙏

File: inference/core/workflows/core_steps/analytics/line_following/v1.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    detections: StepOutputSelector(
+        kind=[
+            OBJECT_DETECTION_PREDICTION_KIND,
+            INSTANCE_SEGMENTATION_PREDICTION_KIND,
+        ]
+    ) = Field(  # type: ignore
+        description="Predictions",
+        examples=["$steps.object_detection_model.predictions"],
+    )
+    triggering_anchor: Union[str, WorkflowParameterSelector(kind=[STRING_KIND])] = Field(  # type: ignore
```
[the remark this answers is not in the document]
Comment (shantanubala) [reply inference/core/workflows/core_steps/analytics/line_following/v1.py]: Fixed 👍

File: inference/core/workflows/core_steps/analytics/line_following/v1.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    def get_manifest(cls) -> Type[WorkflowBlockManifest]:
+        return LineFollowingManifest
+
+    def _calculate_frechet_distance(
+        self, path1: np.ndarray, path2: np.ndarray
+    ) -> float:
+        def euclidean_distance(point1, point2):
+            return np.sqrt(np.sum((point1 - point2) ** 2))
+
+        def compute_distance(dist_matrix, i, j, path1, path2):
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/analytics/line_following/v1.py]: just fine if you record loom I would watch and ask follow up questions if I have one, I would be fine if you explain the concept of algorithm and some basic details regarding implementation

[the remark this answers is not in the document]
Comment (shantanubala) [reply inference/core/workflows/core_steps/analytics/line_following/v1.py]: That sounds good I will record a Loom to share tomorrow 👍

File: inference/core/workflows/core_steps/analytics/path_deviation/v1.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+            anchor_point = anchor_points[i]
+            if tracker_id not in self._object_paths[video_id]:
+                self._object_paths[video_id][tracker_id] = []
+            self._object_paths[video_id][tracker_id].append(anchor_point)
+
+            object_path = np.array(self._object_paths[video_id][tracker_id])
+            ref_path = np.array(reference_path)
+
+            frechet_distance = self._calculate_frechet_distance(object_path, ref_path)
+            detection[DETECTIONS_PATH_DEVIATION_PARAM] = np.array(
```
Comment (PawelPeczek-Roboflow) [inline inference/core/workflows/core_steps/analytics/path_deviation/v1.py:125]: @grzegorz-roboflow please ack this

that's change in the direction of extending our "extended metadata" interface for `sv.Detections`. I believe this is ok, yet for this field to be serialised we need to add change [here](https://github.com/roboflow/inference/blob/e7578d5f4729462a16a688df1fa35c885f0bf537/inference/core/workflows/core_steps/common/serializers.py#L29)

**Review by grzegorz-roboflow — APPROVED** [review]

**Pull request merged**